README.md
# Lockpage Full Stack Starter

Deploy a new password protected Next.js app to AWS securely in 10 minutes.

* [Overview](#overview)
* [Getting Started](#getting-started)
  * [Prerequisites](#prerequisites)
  * [Installation](#installation)
  * [Try the lockpage locally](#try-the-lockpage-locally)
  * [Try the variations directly](#try-the-variations-directly)
  * [Use https locally](#use-https-locally)
  * [Deploy to production with AWS CDK & AWS CodePipeline](#deploy-to-production-with-aws-cdk--aws-codepipeline)
  * [Destroy the CDK stack](#destroy-the-cdk-stack)
* [Serving different variations for different passwords](#serving-different-variations-for-different-passwords)
  * [Adding an app variation](#adding-an-app-variation)
* [Development](#development)
  * [Database and data models](#database-and-data-models)
  * [Feature flags](#feature-flags)
* [Testing](#testing)
  * [Jest unit tests](#jest-unit-tests)
  * [Testcafe integration tests](#testcafe-integration-tests)
* [Appendix: Preparing the AWS Prerequisites](#appendix-preparing-the-aws-prerequisites)
  * [|1| Create an AWS account](#1-create-an-aws-account)
  * [|2| Configure the AWS Command Line Interface CLI](#2-configure-the-aws-command-line-interface-cli)
  * [|3| Register a Connection to your GitHub repository](#3-register-a-connection-to-your-github-repository)
  * [|4| Purchase or import a domain name in Route 53 $/year](#4-purchase-or-import-a-domain-name-in-route-53-year)
  * [|5| Create a Route 53 Hosted Zone for your domain $0.50/month](#5-create-a-route-53-hosted-zone-for-your-domain-050month)
  * [|6| Create or import an HTTPS Certificate in AWS Certificate Manager ACM](#6-create-or-import-an-https-certificate-in-aws-certificate-manager-acm)
* [Appendix: Full Stack](#appendix-full-stack)
* [Credits](#credits)

<br />

## Overview

Quickly deploy a new [Next.js](https://nextjs.org) app to AWS while protecting it behind a **"lockpage"**: a configurable password protection page where **different passwords** can serve **different variations** of the app, or different apps entirely.

<p><img src=".readme/lockpage.png" alt="Default Lockpage screenshot" width="800" /></p>

It's secured by a **signed JSON Web Token (JWT)** and **signed cookie**. 

A complete AWS infrastructure is defined by the included **AWS Cloud Development Kit (CDK) stack** (see [Appendix: Full Stack](#appendix-full-stack) for an overview). After secrets configuration, you'll run one command to:

1.  Create **AWS infrastructure**
2.  Deploy your code from GitHub to a **https** url
3.  Configure **continuous deployment** to re-deploy changes automatically

All resources provisioned by the CDK stack **run for free** under the AWS Free Tier (for 12 months). Your costs are the **domain name** and associated **Route 53 Hosted Zone**, which are required to secure your infrastructure behind **HTTPS** (domain names run around *$10/year*; a Hosted Zone is *$0.50/month*).

## Getting Started

### Prerequisites

To try it locally, you will need:

*   `zsh` or `bash`
*   `wget` or `curl`
*   `rename`
*   `node` *(v16.x.x)*
*   `yarn` *(v1.x)*
*   *Docker Engine* installed and running

Later, to deploy to AWS, you will need:

1.  An **AWS account** created
2.  **AWS CLI** configured
3.  A **domain name** purchased in or imported into Amazon Route 53
4.  A Route 53 **Hosted Zone** created for domain name
5.  **Connection to GitHub** account/repo created in AWS Dev Tools
6.  ~~**HTTPS certificate** created or imported in AWS Certificate Manager~~ *(not needed as of 0.1.3)*

(See [Appendix: Preparing the AWS Prerequisites](#appendix-preparing-the-aws-prerequisites) for a detailed walkthrough)

### Installation

1.  `cd` to a project directory
2.  Run the install script via `wget` or `curl`:

    ```sh
    # (optionally replace zsh with bash)

    # wget
    $ zsh <(wget -qO- https://raw.githubusercontent.com/tw-space/lockpage-full-stack-starter/master/scripts/create-lockpage-app.zsh)

    # curl
    $ zsh <(curl -fsSo- https://raw.githubusercontent.com/tw-space/lockpage-full-stack-starter/master/scripts/create-lockpage-app.zsh)
    ```

### Try the lockpage locally

The starter is configured with **two variations** to start, initially called **main** and **guest**. These correspond to the pages stored under `src/pages/_main/` and `src/pages/_guest/` respectively. The correct password entered into the lockpage will serve either the **main** or **guest** variation.

To try this locally:

1.  First, run the setup script and follow any instructions that appear:

    ```sh
    $ yarn dev:setup
    ```

2.  Once all setup steps are complete, build the project and serve it locally in development mode by running:

    ```sh
    $ yarn serve:dev
    ```

3.  Visit http://localhost:3000 in a browser
4.  Try a few inputs
5.  Enter `main secret` to access the **main** app variation
6.  In the same tab, navigate to http://localhost:3000/enter via the address bar
7.  Enter `guest secret` to access the **guest** variation
8.  To stop the custom Express server, it is not enough to stop the foreground process as it is run with `pm2` as a background process. Instead, run:

    ```sh
    $ yarn stop
    ```

### Try the variations directly

For faster development cycles, the development server can bypass the lockpage and serve the desired app variation directly. Additionally, doing so will enable **hot reloading**, so that app code changes reflect in the browser immediately.

To run the **main** variation directly:

1.  Run:

    ```sh
    $ yarn dev      # same as yarn dev:main
    ```

2.  Visit http://localhost:4000

To run the **guest** variation:

1.  Run:

    ```sh
    $ yarn dev:guest
    ```

2.  Visit http://localhost:4000

To stop the hot reloading server, simply stop the foreground process with `Ctrl-C`.

### Use https locally

By default, both development servers will run in *http*, not *https*. The starter simplifies the setup for **local https development**, however, this is true only for the **custom server** (started with `yarn serve:dev` or `yarn start:dev`, *not* the hot reloading server started with `yarn dev`). 

Steps to use https locally:

1.  Run `yarn https:local:setup` and follow any instructions
2.  Run `mkcert -install` if directed to do so in the output from step 1. Enter the *Sudo password* when prompted. 
3.  Run `yarn serve:dev` (or `yarn start:dev` if pages are already built) to serve locally with *https*
4.  Visit the app in the browser at https://localhost:3000

### Deploy to production with AWS CDK & AWS CodePipeline

This starter streamlines your deployment to a fully functional and secure stack hosted on AWS and set up for **Continuous Deployment**. Your `cdk/my-app-cdk/` directory contains *Infrastructure as Code (IaC)*, a complete infrastructure stack written with the **AWS Cloud Development Kit (CDK)**, which defines and configures AWS elements in TypeScript. 

Once your AWS prerequisites are set up, the entire first-time stack and code deployment takes **less than 10 minutes**. All resources created by the stack **run for free** under the AWS Free Tier (which lasts 12 months from account creation) when no other resources are also running (in particular other Load Balancers, EC2, or RDS instances).

Here are the steps:

1.  Change directory to `cdk/my-app-cdk/`
2.  Run `yarn cdk:precheck` or manually rename `RENAME_TO.secrets.js` to `.secrets.js` in `cdk/my-app-cdk/.env/`
3.  Commit & push your new full project directory to a GitHub repo
4.  Ensure the following are prepared for AWS (see [Appendix: Preparing the AWS Prerequisites](#appendix-preparing-the-aws-prerequisites) for a detailed walkthrough):
    1.  An **AWS account** created
    2.  **AWS CLI** configured
    3.  A **domain name** purchased in or imported into Amazon Route 53
    4.  A Route 53 **Hosted Zone** created for domain name
    5.  **Connection to GitHub** account/repo created in AWS Dev Tools
    6.  ~~**HTTPS certificate** created or imported in AWS Certificate Manager~~ *(not needed as of 0.1.3)*
5.  Set the proper values for these in `cdk/my-app-cdk/.env/.secrets.js`
6.  To serve your web app using free, properly signed **HTTPS** encryption:
    1.  Navigate to `cdk/create-certs-cdk/` 
    2.  Set the required `CDK_CERT_` values in `.env/.secrets.js`
    3.  Decide to store values in **SSM Parameter Store** or to hardcode them in a file, then follow the instructions accordingly in `user-data/run-certbot.sh`
    4.  Run `yarn cdk:full` and follow any instructions
    5.  When successfully complete:
        1.  **Destroy** the stack
        2.  Set `useHttpsFromS3` to `'1'` in `cdk/my-app-cdk/.env/.secrets.js` and in `.env/.secrets.js`
7.  Optionally change any project secret defaults in `.env/.secrets.js`
8.  Put all project secrets prefixed with `jwt`, `secret`, or `dbProd` in your AWS SSM Parameter Store per the steps in `.env/.secrets.js`. For example:

    ```sh
    $ aws ssm put-parameter \
      --name '/my-app/prod/jwtSubMain' \
      --value 'jwtSubMain secret' \
      --type 'SecureString'
    ```

9.  From `cdk/my-app-cdk/`, synthesize and deploy your app infrastructure stack with:

    ```sh
    $ yarn cdk:full
    ```

10. When that's complete, go to **CodePipeline console** and wait until the deployment is fully complete
11. When complete, visit your domain name (using **https**) in a browser
12. Push future changes to master to **continuously deploy** your app to production

### Destroy the CDK stack

While all resources run for free under the Free Tier, it's a good practice to keep usage minimal by **regularly destroying** the **CDK stack** when not actively needing the production deployment:

1.  From `cdk/my-app-cdk/`, run:

    ```sh
    $ yarn cdk:destroy
    ```

## Serving different variations for different passwords

Your `src/pages/` directory looks like this:

```sh
├── pages/
│   ├── _app.page.jsx
│   ├── _document.page.jsx
│   ├── _guest/
│   │   ├── index.page.jsx
│   │   └── index.spec.js
│   ├── _main/
│   │   ├── index.page.jsx
│   │   └── index.spec.js
│   └── _middleware.page.js
```

Notice the two directories `_main/` and `_guest/`. Those are the two default **app variations** that are served for **different passwords**.

*   To visit the pages under `_main/`, type in at the lockpage the default password: `main secret`
*   To visit the pages under `_guest/`, type in at the lockpage the default password: `guest secret`

Once authenticated, the pages you have access to will *only* be those in your permitted directory. For example, after authenticating with `main secret`, you will only be served the pages under `_main/` — though they will appear to be served at the root — and you will not be able to access any pages under `_guest/`.

You may freely add an app variation in the following way:

### Adding an app variation

Let's add an app variation called `cinematic`.

1.  First, create a new directory `_cinematic/` inside `src/pages/`
2.  Next, open `_middleware.page.js`, also in `src/pages/`, and change the following:
    1.  Add the path `_cinematic/` to the array `pathBases`:

        ```js
        const pathBases = [
          '/_main',
          '/_guest',
          '/_cinematic',
        ]
        ```
    
    2.  Extend the first `pathname` check with `pathBases[2]`:

        ```js
        if (
          pathname.startsWith(`${pathBases[0]}`)
          || pathname.startsWith(`${pathBases[1]}`)
          || pathname.startsWith(`${pathBases[2]}`)
        ) {
          return res.rewrite('/404')
        }
        ```
    3.  Add a case to the switch statement like the following:

        ```js
        // ...
        case process.env.JWT_SUB_CINEMATIC:
          return res.rewrite(`${pathBases[2]}${pathname}`)
        // ...
        ```

3.  Now let's add that mystery environment variable and one other. Open `.env/.secrets.js`, and add **two environment variables**:

    ```js
    const jwtSubCinematic = 'cinematic' // a unique permissions identifier that no one will see        
    const secretKeyCinematic = '<PASSWORD>' // the password to access the cinematic variation

    module.exports = {
      jwtSubCinematic,
      secretKeyCinematic,
    }
    ```

4.  Open `.env/common.env.js`, and add **two entries**:

    ```js
    const envCommon = {
      JWT_SUB_CINEMATIC: '',
      SECRET_KEY_CINEMATIC: '',
    }
    ```

5.  Open `.env/production.env.js`, and add:

    ```js
    const {
      jwtSubCinematic,
      secretKeyCinematic,
    } = require('./.production.secrets.js')

    const envProduction = {
      JWT_SUB_CINEMATIC: jwtSubCinematic || '',
      SECRET_KEY_CINEMATIC: secretKeyCinematic || '',
    }
    ```

6.  Make those same additions to `.env/development.env.js` and `.env/testing.env.js`
7.  With that complete, open `server/index.js`, find the switch statement, and add the appropriate case:

    ```js
    switch (req.body[process.env.KEY_NAME]) {
      case process.env.SECRET_KEY_MAIN:
        jwtSub = process.env.JWT_SUB_MAIN
        break
      case process.env.SECRET_KEY_GUEST:
        jwtSub = process.env.JWT_SUB_GUEST
        break
      case process.env.SECRET_KEY_CINEMATIC:    // add this case
        jwtSub = process.env.JWT_SUB_CINEMATIC
        break
      default:
        res.status(500).json({ message: Strings.msg500ServerError })
    }
    ```

8.  Open `utils/auth.js`, and similarly extend the patterns that are there. There are **four** in total: **two** which will reference `process.env.SECRET_KEY_CINEMATIC` and **two** which will reference `process.env.JWT_SUB_CINEMATIC`
9.  Finally, put the secrets into **SSM Parameter Store** by running:

    ```sh
    $ aws ssm put-parameter --name "/<your-app>/prod/jwtSubCinematic" --value "<value>" --type "SecureString"
    $ aws ssm put-parameter --name "/<your-app>/prod/secretKeyCinematic" --value "<value>" --type "SecureString"
    ```

10. That's it! Run the server locally with `yarn serve:dev` to try it. Then to test in production, create pages under `_cinematic/`, commit & push your changes to GitHub, wait a few minutes for your changes to deploy, then visit your domain to check the result.

## Test

The starter includes **Jest unit tests** and **Testcafe integration tests**.

### Jest (unit tests)

Run the included **Jest unit tests** with:

```sh
$ yarn jest:all:build   # builds all, starts custom server, then runs all tests
$ yarn jest             # runs all tests except server (same as yarn jest:all:noserver)
$ yarn jest:app         # only runs tests under src/
```

### Testcafe (integration tests)

Testcafe can run **integration tests** in an actual (headless) browser. It supports **several browsers**, including Chrome, Firefox, and Safari.

1.  First, start the server. You can do it one of two ways:

    ```sh
    $ yarn dev          # starts hot reloading src/pages/_main/ at http://localhost:4000
    $ yarn start:test   # requires 'yarn build:all:test' first; runs at https://localhost:3000
    ```

2.  Depending on how you started the server, run `testcafe`:

    ```sh
    $ yarn testcafe       # test against http://localhost:4000
    $ yarn testcafe:main  # test against https://localhost:3000
    ```

### PR Tests

GitHub will automatically run **PR tests** via the included [GitHub Actions](https://docs.github.com/en/actions/learn-github-actions) workflow when creating a new Pull Request into the *master* branch.

## Development

### Database and data models

This starter is configured to run and interface with **PostgreSQL** in local development and in production on an AWS RDS instance.

For **local development**, running `yarn dev` will first run a script that will walk you through any unmet steps for running a *Postgres* database container locally (namely to have *Docker Engine* installed and running).

This starter is preconfigured for **migration-based database development**, where all changes to a database are reflected in a sequence of **migration files**. Migration files are checked into a repository along with feature code, which makes the database effectively "versioned" alongside commits and releases.

The system for running and tracking migrations is enabled by [Flyway](https://flywaydb.org) whose Community version is free to use. As Flyway is Java based, this starter provides scripts to run Flyway commands in a Docker container spun up on demand.

Interacting with the database in client code is achieved through *Knex.js* database connections and *Objection.js* Model objects (which are built on *Knex.js*).

A quick way to understand how all the pieces fit together is with an illustration of a typical workflow of making a change to the database:

1.  You're working on version *0.0.0* of your app. You create a new sql file in `db/migrations` with a title following a very specific format, such as `V0.0.0_0__My_awesome_database.sql` (see [Flyway Migrations](https://flywaydb.org/documentation/concepts/migrations.html) for their naming conventions and requirements).

2.  You write the changes you want to make to your database in *sql*, for example:

    ```sql
    /* V0.0.0_0__My_awesome_database.sql */
    CREATE TABLE IF NOT EXISTS awesome_things (
      awesome_thing_id int GENERATED ALWAYS AS IDENTITY primary key,
      display_name text NOT NULL
    );
    ```

3.  You add an additional sql migration file to insert test content into your database called `V0.0.0_1__TEST_Insert_awesome_things.sql`. Flyway will run migration files **in order** by their version number (in this case, `0.0.0_1` after `0.0.0_0`). (For ideas on how to organize release and test migration files separately, see [Organising your migrations](https://flywaydb.org/blog/organising-your-migrations).

4.  You're ready to create your database. You run `yarn db:dev:setup` (or simply `yarn dev`) to run Flyway's `migrate` command on a new local Postgres database, all with Docker Compose. The script completes successfully and your local database is initialized and ready to go.

5.  (Optional) You want to interact with the local database directly. You make sure `postgresql` is installed, and you run `yarn db:dev:connect:psql` to connect to the local database with `psql`. 

6.  In your Next.js pages and components, you interface with your database primarily through Objection.js Models which you create in the `models/` directory. You create `models/AwesomeThings.js`, which defines a model for your database table per the guidelines at [Objection.js](https://vincit.github.io/objection.js/). You sometimes may interface with the database more directly by importing and using the `db/knex` object.

7.  You write your pages' database queries as much as possible inside `getServerSideProps()` to achieve *server-side rendering*. (See [Data Fetching: getServerSideProps](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props)) 

8.  You run `yarn dev` to interact with your app in the browser. `yarn dev` will also run all migrations found in `db/migrations/` on the locally running Postgres container.

9.  You write tests for your database in `db/test/` and tests for your models in `models/`. These will automatically be discovered when running `yarn jest` and the `yarn jest:all:` variants. You can also run them specifically with `yarn jest:db` and `yarn jest:models`.

10. (Optional) Your database and migration files look good and you're ready to check in your changes. You push your changes to a development branch then create a Pull Request into master. This triggers a [GitHub Actions](https://docs.github.com/en/actions) workflow which sets up all dependencies, including Flyway and Postgres, and ultimately runs `yarn jest:all:prod` to run all discoverable Jest tests.

11. You commit your changes to master. Your app is already deployed to AWS via the included CDK stack and it is running a Free Tier RDS PostgreSQL instance. CodePipeline detects your changes in the repository and starts a CodeDeploy deployment. As part of the deployment scripts, a Flyway container is spun up to run `migrate` on your production database. Your database and app code update successfully and your changes are live.

### Feature flags

This starter includes basic support for **feature flags** to support [trunk-based development](https://trunkbaseddevelopment.com). 

To illustrate, when beginning a new feature, you can:

1.  Add a new flag in `.env/development.flags.js` with name `FLAG_NEW_FEATURE` and set its value to `'on'`:

    ```js
    // .env/development.flags.js
    const flagsDevelopment = {
      FLAG_NEW_FEATURE: 'on',
    }
    ```

    *Note:* Flags' environment variable names must be in **snake uppercase** format and begin with `'FLAG_'`.
    
2.  In app code, write new implementation logic that runs in development mode but **not in production** by surrounding it like in this example:

    ```js
    import * as flag from '../utils/code-flags'

    if (flag.isEnabled('newFeature')) {
      /* Only runs if process.env.FLAG_NEW_FEATURE is set to 'on' */
    }
    else { /*...*/ }
    ```

3.  Alternatively when **testing variants**, use `getVariant()` to check a flag's value directly:

    ```js
    import * as flag from '../utils/code-flags'
    
    if (flag.getVariant('featureWithVariant') === 'blue') {
      /* Only runs if process.env.FLAG_FEATURE_WITH_VARIANT has value 'blue' */
    } else if (flag.getVariant('featureWithVariant') === 'green') {
      /* Only runs if process.env.FLAG_FEATURE_WITH_VARIANT has value 'green' */
    }
    else { /*...*/ }
    ```
    
4.  Commit this code into the trunk knowing it **won't affect production code**.
5.  When ready to **enable in production**, add the appropriate flag in `.env/production.flags.js` and push to remote master:

    ```js
    // .env/production.flags.js
    const flagsProduction = {
      FLAG_NEW_FEATURE: 'on',
    }
    ```

6.  Finally, when the feature is demonstrated to work in production and all is well, remember to **remove this flag's code** to keep things tidy.

## Appendix: Preparing the AWS Prerequisites

To configure your app for deployment to AWS, you will need to provide these six values in `cdk/my-app-cdk/.env/.secrets.js`:

```js
const cdkGitHubConnectionArn = ''
const cdkGitHubOwner = ''
const cdkGitHubRepo = ''
const cdkGitHubRepoBranch = ''
const cdkHostedZoneId = ''
const cdkHostname = ''
```

You will know these values after these **six steps**:

### `|1|` Create an AWS account

Here's an overview of the first steps: [Prerequisites to use the AWS CLI version 2 | AWS](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-prereqs.html)

1.  Go to [AWS sign up](https://portal.aws.amazon.com/billing/signup) and create an account for free.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.1.5 — Fixes to Create App Script (2022-04-05)

This release adds **fixes** to the **create-lockpage-app** script.

### Adds

* Conversion of `MyApp` to pascal cased transformation of app name in *create-lockpage-app* script
* Third party notice for this starter in `THIRD_PARTY_NOTICES.md`
* `CHANGELOG_NEW.md` with empty changelog which is renamed to `CHANGELOG.md` at create
* Automatic copying of `RENAME_TO.secrets.js` to `.secrets.js` files
* Populates `rootPwd` in `.secrets.js`

### Changes

* Defaults `version` to `0.0.0` in `package.json` after create
* Defaults `license` to `UNLICENSED` in `package.json` after create
* Removes `LICENSE` file after create
* Clears `description` field in `package.json` after create
* Removes `repository` info from `package.json` after create
* Improves `README`
* Improves preparations for `create-certs-cdk` stack after create

## 0.1.4 — CI Pipeline (2022-03-11)

This release adds **Continuous Integration** into the CodePipeline.

### Adds

* Generated unique S3 artifact bucket name to be reused by subsequent runs of CodePipeline
* Fixes to create-lockpage-app script
* Scripts to simplify project setup (dev:setup, db:dev:setup)
* Simplified https local setup via https:local:setup script
* CodePipeline dependency on ec2 creation in cdk script

## 0.1.3 — Database (2022-03-03)

The focus for this release is adding a **database**.

### Adds

* Amazon RDS PostgreSQL instance (t2.micro, free tier) in vpc via app CDK stack
* Database migrations run with Flyway via Docker Compose
* Docker Compose scripts which spin up Postgres and Flyway containers and run Flyway migrate on SQL migration files to populate database (see `yarn db:` scripts). Configurations adapted for local development, PR tests in GitHub Actions, and deployment to AWS via the CDK stack.
* Knex.js library for low-level access to database
* Objection.js library, an abstraction layer built on Knex.js, for building and interacting with data models
* Ability to connect to RDS PostgreSQL instance remotely via port forwarding through the EC2 instance with *socat* (use `yarn db:prod:connect:psql`)
* Minimal front-end example under *_main* which pulls "Famous People" from database with Server-Side Rendering (images fetched client-side in prototype style; *later:* store images in S3)
* Database tests (`yarn jest:db`)
* Model tests (`yarn jest:models`)
* Support for also running database and model PR tests in GitHub Actions workflow using Docker Compose when creating a Pull Request into master
* Separate standalone CDK stack which spins up a temporary EC2 instance, associates it with a Route 53 alias record for a domain name you own, requests a free HTTPS (TLS) certificate and key from Let's Encrypt, and stores them in a private S3 bucket. Also adds script prechecks when synthesizing the stack to give guidance and avoid duplications. Run `yarn cdk:full` in *cdk/create-certs* to get started.
* Configurability in the app CDK EC2 instance to use the certificate and key stored in S3 to serve via properly signed HTTPS (set `CDK_` environment variables in `cdk/my-app-cdk/.env/.secrets.js`)
* Script to check for and give guidance on setting up local HTTPS when testing and serving the full built project locally including the lockpage (see `yarn https:local:setup`)
* Better logging for EC2 user data and CodeDeploy scripts. Find user data logs in */var/log* and CodeDeploy logs in *~/log* on the EC2 instance.

### Changes

* EC2 Auto Scaling group and load balancer changed to EC2 instance in app cdk (to ensure EC2 and RDS single instances are placed in same availability zone, thus avoiding data transfer costs across AZs)

## 0.1.2 — Feature Flags (2022-02-12)

This release adds support for **feature flags**.

### Adds

* Helpers for **feature flags**. Details in *README*.

## 0.1.1 — PR tests (2022-02-12)

This release adds support for **PR tests** in GitHub via a GitHub Actions workflow.

### Adds

* GitHub Actions workflow that triggers PR tests to run on creating a new Pull Request into *master*

## 0.1.0 — First Release (2022-01-26)

First public release of this starter. See *README* for details.

### Adds

* AWS CDK stack which quickly deploys your project to a domain name you own. The stack creates:
  * Elastic Application Load Balancer configured to listen with an HTTPS certificate you own and store in AWS Certificate Manager (free to create)
  * EC2 Auto Scaling Group configured to create one instance which only allows inbound traffic from the load balancer
  * Route 53 Alias Record which directs traffic for your owned domain name to the load balancer
  * CodePipeline which Continuously Deploys changes to your GitHub repository via CodeDeploy
* Next.js starter protected behind a "lockpage" — a static password protection page interacting with a custom Express server and Next.js middleware which can authorize access to different variants of your project for different passwords. Uses JSON Web Tokens and secure cookies for authorization. 
* Comprehensive *Jest* unit tests for client and server (uses *Testing Library*, *Mock Service Worker*, and *supertest*)
* Support for styling with *Stitches*
* Script to create new projects from this starter template with a given app name
* Lots of *yarn* scripts in *package.json* to automate tasks and configurations


server/index.js
//
// server / index.js
//
const fs = require('fs')
const https = require('https')
const path = require('path')

const cookieParser = require('cookie-parser')
const express = require('express')
const helmet = require('helmet')
const jwt = require('jsonwebtoken')
const morgan = require('morgan')
const nextStart = require('next')
// const enforce = require('express-sslify')

const { Strings } = require('../resources/Strings')

const port = parseInt(process.env.PORT, 10) || 3000
const isDev = process.env.TRUE_ENV !== 'production'
const isPRTest = process.env.TRUE_ENV === 'test-PR'
const nextApp = nextStart({ isDev, customServer: true, hostname: 'localhost', port })
const handle = nextApp.getRequestHandler()

const lockpageStaticDir = path.resolve(__dirname, `../${process.env.LOCKPAGE_DIR}`)
const loginPath = process.env.LOGIN_PATH
const useHttpsLocal = process.env.USE_HTTPS_LOCAL === '1'
const useHttpsFromS3 = process.env.USE_HTTPS_FROM_S3 === '1'

//
// Configure and start the custom server
//

nextApp
  .prepare()
  .then(() => {

    const app = express()

    // Configure secure headers with helmet, unless using http local or in PR test
    if (!((isDev && !useHttpsLocal) || isPRTest)) {
      app.use(setHelmet())
    }

    // Logging (for development for now)
    if (isDev) {
      app.use(morgan('dev'))
    }

    // Enforce HTTPS in production (done differently in dev)
    // trustProtoHeader needed for load balancer (not working)
    // if (!isDev) {
    //   app.use(enforce.HTTPS({ trustProtoHeader: true }))
    // }

    // Enable reading form data
    app.use(express.json())
    app.use(express.urlencoded({ extended: true }))

    // Enable signed cookies
    app.use(cookieParser(process.env.SECRET_COOKIE))

    // Serve only the locksite before authentication and set with virtual path prefix
    app.use(
      loginPath,
      express.static(lockpageStaticDir),
    )

    // Authenticate with keyphrase
    app.post('/unlock', unlockWithKey)

    // Ensure authenticated before serving the main site
    app.use(authenticateJWT)

    // Authenticated; now pass to Next handler
    app.get('*', (req, res) => handle(req, res))

    if (!isDev && useHttpsFromS3 && !isPRTest) {
      https
        .createServer(
          {
            cert: fs.readFileSync('../.certificates/fullchain.pem'),
            key: fs.readFileSync('../.certificates/privkey.pem'),
          },
          app,
        )
        .listen(port, (err) => {
          if (err) throw err
          // eslint-disable-next-line no-console
          console.log(`Node prod server (https): listening on port ${port}`)
        })
    } else if (isDev && useHttpsLocal) {
      // Use localhost https when set in development and test
      https
        .createServer(
          {
            key: fs.readFileSync('./server/localhost.key'),
            cert: fs.readFileSync('./server/localhost.crt'),
          },
          app,
        )
        .listen(port, (err) => {
          if (err) throw err
          // eslint-disable-next-line no-console
          console.log(`Node dev server (https): listening on port ${port}`)
        })
    } else {
      app.listen(port, (err) => {
        if (err) throw err
        // eslint-disable-next-line no-console
        console.log(
          `Node ${
            // eslint-disable-next-line
            isDev ? 'dev server' : 'process ' + process.pid
          }: listening on port ${port}`,
        )
      })
    }
  })
  .catch((err) => {
    console.error(err) // eslint-disable-line no-console
  })

//
// Functions
//

function setHelmet() {
  return helmet({
    contentSecurityPolicy: {
      directives: {
        defaultSrc: [`'self'`],
        baseUri: [`'self'`],
        connectSrc: [`'self'`],
        fontSrc: [`'self'`, `data:`],
        formAction: [`'self'`],
        frameAncestors: [`'none'`],
        frameSrc: [`'none'`],
        imgSrc: [`'self'`, `data:`],
        mediaSrc: [`'none'`],
        objectSrc: [`'none'`],
        sandbox: ['allow-forms', 'allow-scripts', 'allow-same-origin'],
        styleSrc: [`'self'`, `'unsafe-inline'`],
        scriptSrc: [`'self'`],
        scriptSrcAttr: [`'none'`],
        blockAllMixedContent: [],
        upgradeInsecureRequests: [],
      },
    },
    referrerPolicy: {
      policy: ['no-referrer', 'strict-origin-when-cross-origin'],
    },
  })
}

function bearsKey(req) {
  if (
    !(
      process.env.KEY_NAME
      && process.env.SECRET_KEY_MAIN
      && process.env.SECRET_KEY_GUEST
    )
  ) {
    return false
  }
  return (
    req.body[process.env.KEY_NAME] === process.env.SECRET_KEY_MAIN
    || req.body[process.env.KEY_NAME] === process.env.SECRET_KEY_GUEST
  )
}

function unlockWithKey(req, res) {
  if (bearsKey(req)) {
    let jwtSub
    switch (req.body[process.env.KEY_NAME]) {
      case process.env.SECRET_KEY_MAIN:
        jwtSub = process.env.JWT_SUB_MAIN
        break
      case process.env.SECRET_KEY_GUEST:
        jwtSub = process.env.JWT_SUB_GUEST
        break
      default:
        res.status(500).send(Strings.msg500ServerError)
    }
    const signedJWT = jwt.sign({}, process.env.SECRET_JWT, {
      issuer: process.env.JWT_ISS,
      audience: process.env.JWT_AUD,
      expiresIn: process.env.JWT_EXP_IN,
      subject: jwtSub,
    })
    if (!signedJWT) {
      res.status(500).send(Strings.msg500ServerError)
    } else {
      res.cookie(process.env.JWT_NAME, signedJWT, {
        maxAge: process.env.JWT_EXP_IN,
        secure: !((isDev && !useHttpsLocal) || isPRTest),
        httpOnly: true,
        sameSite: 'Strict',
        signed: true,
      })
      // redirect to '../' since lockpage is served at login path
      res.status(200).json({ redirect: '../' })
    }
  } else {
    res.status(401).send(Strings.msg401WrongKey)
  }
}

function authenticateJWT(req, res, next) {
  // Confirm correct server environment variables
  if (
    !(
      process.env.JWT_ALG
      && process.env.JWT_AUD
      && process.env.JWT_EXP_IN
      && process.env.JWT_ISS
      && process.env.JWT_NAME
      && process.env.JWT_SUB_MAIN
      && process.env.JWT_SUB_GUEST
      && process.env.SECRET_JWT
      && process.env.LOGIN_PATH
    )
  ) {
    res.status(500).send(Strings.msg500ServerError)

  // Verify jwt is expected
  } else {
    const requestJWT = req.signedCookies[process.env.JWT_NAME] || ''
    try {
      if (!requestJWT) {
        // no cookie, redirect
        throw 303 // eslint-disable-line no-throw-literal
      } else {
        const jwtDecoded = jwt.decode(requestJWT)
        const decodedSub = jwtDecoded.sub
        if (
          !(
            decodedSub === process.env.JWT_SUB_MAIN
            || decodedSub === process.env.JWT_SUB_GUEST
          )
        ) {
          throw 'invalid jwt sub' // eslint-disable-line no-throw-literal
        }
        jwt.verify(requestJWT, process.env.SECRET_JWT, {
          algorithms: [process.env.JWT_ALG],
          issuer: process.env.JWT_ISS,
          audience: process.env.JWT_AUD,
          maxAge: process.env.JWT_EXP_IN,
          subject: decodedSub,
        }, (error) => {
          if (error) {
            throw 'jwt verify failed' // eslint-disable-line no-throw-literal
          }
        })
        next()
      }
    } catch (err) {
      switch (err) {
        case 303:
        case 'invalid jwt sub':
        case 'jwt verify failed':
        default:
          res.redirect(303, process.env.LOGIN_PATH)
          break
      }
    }
  }
}
